Move device plugin related code to pkg/kubelet/cm/deviceplugin/.

jiayingz opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:
Move device plugin related code to pkg/kubelet/cm/deviceplugin/.
Inspired by PR 53258 proposed by @dolftax. The code structure looks much cleaner this way and it also makes future code refactoring easier.
While moving device_plugin_handler_test.go from pkg/kubelet/cm/ to pkg/kubelet/cm/deviceplugin/, we can no longer uses cm in its tests because that would cause a cycle dependency. To solve this problem, I moved the main cm GetResources functionality as well as part of the current device plugin handler Allocate functionality into a new device plugin handler function, GetDeviceRunContainerOptions(). This refactoring is also needed by another PR 51895 that moves device allocation into admission phase. Now device plugin handler Allocate() first checks whether there is cached device runtime state and only issues Allocate grpc call if there is no cached state available. The new GetDeviceRunContainerOptions() function simply returns device runtime config from the cached state. To support this change, extended the podDevices struct and checkpoint data structure with device runtime state.

We expect to clean up device plugin interface in follow up PRs. Currently both device plugin handler interface and manager interface are public. This is no longer necessary after this PR given that their code now live in the same package. We should clean up unnecessary public interfaces and functions, and/or consider to merge handler and manager code into a single layer.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:
Part of https://github.com/kubernetes/kubernetes/issues/53170 but not fixes it.

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
```

jiayingz commented:
/release-note-none

jiayingz commented:
/assign @RenaudWasTaken @vishh @dolftax

RenaudWasTaken commented:
The only thing that "bugs" me is that @vishh wanted to have a separate package that does not depend on kubernetes.

IMHO as long as we keep the manager interface this change is good!

lichuqiang commented:
Wondering if you could separate the change in **pod_devices** and merge that part first, I think it may have impact on many upcoming PRs.

lichuqiang commented on pkg/kubelet/cm/deviceplugin/device_plugin_handler.go:
```diff
@@ -29,79 +29,36 @@ import (
+	// for the passed-in <pod, container> and returns its DeviceRunContainerOptions
+	// for the found one. An empty struct is returned in case no cached state is found.
+	GetDeviceRunContainerOptions(pod *v1.Pod, container *v1.Container) DeviceRunContainerOptions
 }
```
It would be a bit tricky to see whether we get real containerDevices or not by returning a DeviceRunContainerOptions item. I would prefer to return a reference here.
In addition,  transmission for structs may have impact on performance

tengqm commented:
/cc

jiayingz commented:
/retest

RenaudWasTaken commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -618,71 +619,22 @@ func (cm *containerManagerImpl) GetResources(pod *v1.Pod, container *v1.Containe
+	if devOpts == nil {
+		return opts, nil
+	}
+	if len(devOpts.Devices) > 0 {
```
Nit: if is unneeded appending nil lists to nil lists returns nil

RenaudWasTaken commented:
Looking into this a bit more I'm very confused by this PR, it seems to be doing a lot more than just moving code from one place to another.

Moving the files has the unintended consequence that it prevents looking at the code diff correctly and makes this whole PR a lot more complicated to read.

Can you split this PR into multiple commits ? This would make it a lot more easier to review.

RenaudWasTaken commented:
> While moving device_plugin_handler_test.go from pkg/kubelet/cm/ to pkg/kubelet/cm/deviceplugin/, we can no longer uses cm in its tests because that would cause a cycle dependency. 

The mentioned test seems to not be a unit test which would explain why you had so many problems trying to move it.
That test is testing a method of `containerManagerImpl` in `device_plugin_handler` which should have been unit tested in `container_manager_linux_test.go` with a stub `DevicePluginHandler`.
This would have allowed to send and test different combinations of `AllocationResponses` to `GetResources`.

> To solve this problem, I moved the main cm GetResources functionality as well as part of the current device plugin handler Allocate functionality into a new device plugin handler function, GetDeviceRunContainerOptions(). 

Can we break this change into another PR?
As mentioned before the whole PR is not very reviewer friendly and the introduced changes have a pretty important scope (checkpoint + new struct + new tests).

The mentioned test is currently disabled so we can just comment it out and re-introduce it in the new PR.

> This refactoring is also needed by another PR #51895 that moves device allocation into admission phase.

Moving this code into another PR shouldn't impact PR #51895 anymore.
Anyways you also mentioned that we needed coordination, I believe we should wait until we have a definite architecture design document before merging anything.

> Now device plugin handler Allocate() first checks whether there is cached device runtime state and only issues Allocate grpc call if there is no cached state available. The new GetDeviceRunContainerOptions() function simply returns device runtime config from the cached state. To support this change, extended the podDevices struct and checkpoint data structure with device runtime state.

See my comment about this completely widening the scope of this PR

vishh commented on pkg/kubelet/cm/deviceplugin/device_plugin_handler.go:
```diff
@@ -0,0 +1,289 @@
+		return nil, nil
+	}
+	if needed < 0 {
+		// Pod is downsizing its resource requests. We need to make Allocate rpc call
```
A pod's resource cannot be updated once admitted by the API server. I'd much rather fail loudly in this case (not crashloop) than handle it.

vishh commented on pkg/kubelet/cm/deviceplugin/device_plugin_handler.go:
```diff
@@ -0,0 +1,289 @@
+		}
+		allocDevices, err := h.devicesToAllocate(podUID, contName, resource, needed)
+		if err != nil {
+			return err
```
Will devices get stranded if Allocate fails? IIUC, devices are removed from Allocatable pool prior to actual allocation.

vishh replied:
nvm. I see the note on `updateAllocatedDevices` below

vishh commented on pkg/kubelet/cm/deviceplugin/device_plugin_handler.go:
```diff
@@ -0,0 +1,289 @@
+		// we have X resource A and Y resource B in total, and two containers, container1
+		// and container2 both require X resource A and Y resource B. Both allocation
+		// requests may fail if we serve them in mixed order.
+		// TODO: may revisit this part later if we see inefficient resource allocation
```
This can result in a race - Imagine a pod getting rejected by the kubelet because it temporarily stranded resources that it has previously allocated to a pod it had rejected because Allocation failed.

vishh replied:
May be run an update before every Allocation and ensure that the update is cheap?

vishh reviewed (approved):
/lgtm
/approve

jiayingz reviewed (commented):
Thanks a lot for the reviews! PTAL.

RenaudWasTaken reviewed (changes requested):
Would like to see this PR at least broken down into multiple commits.
This would allow us to review much more confidently this PR based on a diff rather than try to find bugs in a +701 −447 commit.

If not broken into multiple PRs whose scope are:
- Only to move the code to  pkg/kubelet/cm/deviceplugin/ as it is defined in the title
- Change the behavior of cm

jiayingz commented:
@RenaudWasTaken Thanks a lot for your suggestions! I did consider to break this PR into multiple ones and consulted with @vishh yesterday. He suggested to keep the PR as is, given that we are already in the middle of review and this PR is needed for some ongoing PRs. I have described the scope of the change in both PR and commit descriptions. Hopefully that will help future retrospection. Most of the changes included in this PR are code move around, although unfortunately git is not smart enough to recognize that. I generally don't like to have a commit that leaves tests or build in broken state, but I see your point that this could help speed up code review. I will keep that in mind next time. As of avoiding introducing bugs through this PR change, I have made sure both unit tests and e2e tests continued to pass with the PR.

vishh commented on pkg/kubelet/cm/deviceplugin/device_plugin_handler.go:
```diff
@@ -0,0 +1,290 @@
+		// requests may fail if we serve them in mixed order.
+		// TODO: may revisit this part later if we see inefficient resource allocation
+		// in real use as the result of this.
+		resp, err := h.devicePluginManager.Allocate(resource, allocDevices.UnsortedList())
```
One more consideration for the future: Why not allocate multiple resources in parallel? Even if it's serial, kubelet needs to deal with failed allocations (currently handled via `updateAllocations` method).

jiayingz replied:
Right now we don't expect it is a common case that a container require resources from multiple device plugins. If that happens in the future, I agree we should consider this. Added a comment about this.

vishh reviewed (approved):
/lgtm
/approve

RenaudWasTaken commented:
> although unfortunately git is not smart enough to recognize that. 

Which is why I suggested to break this PR in multiple commits.
Every commit does not need to compile just the last one.

> Most of the changes included in this PR are code move around,  I generally don't like to have a commit that leaves tests or build in broken state, but I see your point that this could help speed up code review. I will keep that in mind next time. As of avoiding introducing bugs through this PR change, I have made sure both unit tests and e2e tests continued to pass with the PR.

It's not even about speeding code review. We can't reasonably expect people to review 1000 lines of changes in a single commit.

Can we at least discuss this in tomorrow's meeting before merging this PR ?

jiayingz committed:
Move device plugin related files under pkg/kubelet/cm/deviceplugin/.

jiayingz committed:
Device plugin code refactoring to cope with file move.

While moving device_plugin_handler_test.go from pkg/kubelet/cm/ to
pkg/kubelet/cm/deviceplugin/, we can no longer uses cm in its tests
because that would cause a cycle dependency. To solve this problem,
I moved the main cm GetResources functionality as well as part of the
current device plugin handler Allocate functionality into a new device
plugin handler function, GetDeviceRunContainerOptions(). This
refactoring is also needed by another PR 51895 that moves device
allocation into admission phase. Now device plugin handler Allocate()
first checks whether there is cached device runtime state and only
issues Allocate grpc call if there is no cached state available.
The new GetDeviceRunContainerOptions() function simply returns device
runtime config from the cached state. To support this change, extended the
podDevices struct and checkpoint data structure with device runtime state.

jiayingz committed:
Move podDevices code into a separate file.

RenaudWasTaken commented on pkg/kubelet/cm/deviceplugin/device_plugin_handler.go:
```diff
@@ -43,27 +50,17 @@ func (pdev podDevices) pods() sets.String {
 	}
 	return ret
 }
-
```
nit: keep new line :)

jiayingz replied:
done.

RenaudWasTaken commented on pkg/kubelet/cm/deviceplugin/device_plugin_handler.go:
```diff
@@ -151,50 +313,89 @@ func NewDevicePluginHandlerImpl(updateCapacityFunc func(v1.ResourceList)) (*Devi
 		glog.V(3).Infof("needs %d %s", needed, resource)
-		_, registeredResource := h.allDevices[resource]
-		if !registeredResource || needed == 0 {
+		if _, registeredResource := h.allDevices[resource]; !registeredResource {
```
Looks like this won't continue if needed = 0

jiayingz replied:
that part now moves to devicesToAllocate().

RenaudWasTaken commented on pkg/kubelet/cm/deviceplugin/device_plugin_handler.go:
```diff
@@ -151,50 +313,89 @@ func NewDevicePluginHandlerImpl(updateCapacityFunc func(v1.ResourceList)) (*Devi
-			glog.V(3).Infof("Found pre-allocated devices for resource %s container %q in Pod %q: %v", resource, container.Name, pod.UID, devices.List())
-			needed = needed - devices.Len()
+		// Updates allocatedDevices to garbage collect any stranded resources
+		// before doing the device plugin allocation.
```
Question for the future: How do we expect to report used / unused devices ?

With the current model if devices get stranded they would be advertised as taken right ?
In that case wouldn't a model that cleans up stranded devices be a better one ?

jiayingz replied:
Good point, I added the code to restore h.allocatedDevices upon allocation failures.

RenaudWasTaken commented on pkg/kubelet/cm/deviceplugin/device_plugin_handler.go:
```diff
@@ -205,74 +406,73 @@ func (h *DevicePluginHandlerImpl) Allocate(pod *v1.Pod, container *v1.Container,
+		// Update internal cached podDevices state.
+		h.Lock()
+		if oldDevices := h.podDevices.containerDevices(podUID, contName, resource); oldDevices != nil {
+			freeDevices := oldDevices.Difference(allocDevices)
```
I might be understanding this wrong but this part of the code seems to suggest that you might not get the same devices between reboots right ?

Wouldn't we want a container to keep the exact same devices between reboots ?

jiayingz replied:
Good point. I added this part when I thought we need to support resource downsize and upsize, but then learned from Vish that a pod's resource requirements can't change as of today. Simplified this part.

RenaudWasTaken commented on pkg/kubelet/cm/deviceplugin/device_plugin_handler.go:
```diff
@@ -205,74 +406,73 @@ func (h *DevicePluginHandlerImpl) Allocate(pod *v1.Pod, container *v1.Container,
 // Checkpoints device to container allocation information to disk.
-func (h *DevicePluginHandlerImpl) writeCheckpoint() error {
+func (h *HandlerImpl) writeCheckpoint() error {
+	h.Lock()
```
Same comment about Big lock here

RenaudWasTaken commented on pkg/kubelet/cm/deviceplugin/device_plugin_handler.go:
```diff
@@ -205,74 +406,73 @@ func (h *DevicePluginHandlerImpl) Allocate(pod *v1.Pod, container *v1.Container,
 // h.allocatedDevices accordingly.
-func (h *DevicePluginHandlerImpl) readCheckpoint() error {
+func (h *HandlerImpl) readCheckpoint() error {
+	h.Lock()
```
Same comment about Big lock here

RenaudWasTaken commented on pkg/kubelet/cm/deviceplugin/device_plugin_handler.go:
```diff
@@ -205,74 +406,73 @@ func (h *DevicePluginHandlerImpl) Allocate(pod *v1.Pod, container *v1.Container,
+// for the passed-in <pod, container> and returns its DeviceRunContainerOptions
+// for the found one. An empty struct is returned in case no cached state is found.
+func (h *HandlerImpl) GetDeviceRunContainerOptions(pod *v1.Pod, container *v1.Container) *DeviceRunContainerOptions {
+	h.Lock()
```
I'm a bit surprised by the usage of the Mutex here. Usually you would want to keep critical section to a minimum.
We are more in a big lock situation here.

jiayingz replied:
I think this is needed to synchronize read/write access on h.podDevices. Also this part shouldn't be expensive.

RenaudWasTaken commented on pkg/kubelet/cm/deviceplugin/device_plugin_handler.go:
```diff
@@ -205,74 +406,73 @@ func (h *DevicePluginHandlerImpl) Allocate(pod *v1.Pod, container *v1.Container,
 		return err
 	}
-	return ioutil.WriteFile(filepath, dataJson, 0644)
+	return ioutil.WriteFile(filepath, dataJSON, 0644)
```
This needs to be out of the critical section. Performance is going to be very impacted

RenaudWasTaken reviewed (changes requested):
The locking strategy doesn't seem like a good idea.
Critical sections are huge.

We should at least address file write in a critical section in this PR.

We can address the complex locking mechanism in future PRs (discussion could happen in the design document).

The complex locking mechanism is not going to be an easy problem to solve because it relates to the global architecture of the device plugin

jiayingz commented:
/retest

vishh commented:
I feel this PR is a net positive from where the code is today. Do you agree @RenaudWasTaken ? If yes, then let's merge this PR and continue improving the code?

RenaudWasTaken commented:
Agreed :)

vishh commented:
/lgtm
/approve
